src/Http/Http/src/Features/RequestBodyPipeFeature.cs

```diff
@@ -31,19 +34,31 @@ public PipeReader Reader
                 }
 
                 if (_internalPipeReader == null ||
-                    !object.ReferenceEquals(_internalPipeReader.InnerStream, _context.Request.Body))
+                    !ReferenceEquals(_streamInstanceWhenWrapped, _context.Request.Body))
                 {
-                    _internalPipeReader = new StreamPipeReader(_context.Request.Body);
-                    _context.Response.RegisterForDispose(_internalPipeReader);
+                    _streamInstanceWhenWrapped = _context.Request.Body;
+                    _internalPipeReader = PipeReader.Create(_context.Request.Body);
+
+                    _context.Response.OnCompleted((self) => ((RequestBodyPipeFeature)self).Complete(), this);
                 }
 
                 return _internalPipeReader;
             }
             set
             {
                 _userSetPipeReader = value ?? throw new ArgumentNullException(nameof(value));
+                // REVIEW: Should we do this (vvvv) or just get rid of the setter?
```

Comment:
Thoughts?

Reply:
I think we should keep the setter. Kestrel needs to be able to set the optimized pipe writer somehow.

Reply:
Can't Kestrel just implement the feature directly though? It doesn't need the setter to provide an optimized pipe.

Reply:
Yea why would it need the setter

Reply:
Should we just remove the setter entirely then? In the original plan we were keeping it (but making the property on `HttpRequest`/`HttpResponse` read-only). You can always replace the feature entirely so if we make the feature read-only it seems fine to me.

What's the contract here regarding completion? My suggestion is this:

1. If you provide this feature yourself, you complete the pipe sometime
1. If we provide this default feature implementation and someone accesses the pipe, we'll queue an `OnComplete` handler to complete the wrapper
1. If you want to "wrap" or "chain" a pipe, you replace the feature entirely and the replaced pipe will be completed (as per the previous two clauses).

Reply:
Okay, I buy that this feature can be readonly.

I think your plan is spot on.

Reply:
![the camera slowly pans back and a man nods lightly](https://media.giphy.com/media/v2A81Lm2MglLa/giphy.gif)